### PR TITLE
add c ffi so people can call casbin-rs from c code

### PR DESCRIPTION
**Pull request by xcaptain**
Writing a pure C version casbin is quite difficult, but using rust ffi can be very easy. This commit adds C-bindings, so C/C++ programmer can easily integrate casbin into their projects.

### CODE REVIEW

**Comment by xcaptain** [comment]
A simple link demo

图片

File: Cargo.toml [hunk 1]
```diff
@@ -24,6 +24,7 @@ async-std = { version = "1.5.0", optional = true }
 async-trait = "0.1.24"
 log = "0.4.8"
 tokio = { version = "0.2.11", optional = true, default-features = false }
+futures = "0.3"
```
Comment (DevinR528) [inline Cargo.toml]: I hope its ok to give some feedback. @GopherJ told me not to add `futures` to `[dev-dependencies]` so what I did instead was [this](https://github.com/casbin/casbin-rs/blob/master/benches/casbin.rs#L4-L18). If this is a C compatibility thing ignore me, feel free to ignore me either way :smile:!

Comment (xcaptain) [reply Cargo.toml]: Yes, you are right, no need to add `futures` here, we can use `async_std::task::block_on` instead. I will make a change.

**Comment by GopherJ** [comment]
I think we should continue from this PR and make it possible to call rust from c/c++, it's a great idea!

**Comment by hsluoyz** [comment]
I have a question. If this idea (PR) works, do we still need to develop: https://github.com/casbin/casbin-cpp ?

**Comment by DevinR528** [comment]
I bet it could be done, this was a really good idea @xcaptain!! Here's a few links I found while looking around [cbindgen](https://github.com/eqrion/cbindgen/blob/master/docs.md) [more helpful info](https://karroffel.gitlab.io/post/2019-05-15-rust/)

All of the C binding stuff could be in a mod that is conditionally compiled with a conditional `build.rs` file, according to the second link it is possible using the [cpp](https://github.com/mystor/rust-cpp) crate to take the cdylib and generate cpp bindings.

**Comment by xcaptain** [comment]
@DevinR528 I did some experiments with `cbindgen` but sadly it doesn't work, because we have used

```rust
Box<dyn Adapter>
// or
*mut dyn Adapter
```

According to cbindgen's [doc](https://github.com/eqrion/cbindgen/blob/master/docs.md)

> cbindgen also cannot support wide pointers like &dyn Trait or &[T], as their layout and ABI is not guaranteed. In the case of slices you can at least decompose them into a pointer and length, and reconstruct them with slice::from_raw_parts.

but a hand written 

```rust
pub extern "C" fn new_adapter(dsn: *const c_char) -> Box<dyn Adapter>
```

can be converted to

```C
extern struct Adapter *new_adapter(char *dsn);
```

It's a little tricky but we were lucky it worked.

I also tried FFI diesel-adapter and again we were lucky, it worked.

Later I will do some experiments on multi-thread scenarios to see if rust's `RWLock` works in C code, if every thing goes well then I will start porting more APIs to `ffi/c.rs`.

As for the `cpp` crate, it used to embed C++ code into rust project, seems no relation with this pull request. I don't think a C++ binding is needed, because a C binding is enough, also write a well structured C++ binding may be hard.

**Comment by divy9881** [comment]
Hello everyone, this will be a fruitful feature for rust users where they do not require a authorization engine as its C/C++ client project's main requirement but as a RFC(Remote function call) interface like requirement where they do not need to use the whole heavy-weight set-up for C++. Obviously this engine won't be as fast as the main C/C++ library but can be easy to plugin in projects.

**Comment by hsluoyz** [comment]
Any update?

**Comment by hsluoyz** [comment]
Closed as stale

**Pull request closed without merging**